tests/subsys/net/lib/nrf_cloud/nrf_cloud_test/src/nrf_cloud_test.c

```diff
@@ -0,0 +1,609 @@
+	nrf_cloud_fota_fmfu_dev_set_fake.custom_fake = nrf_cloud_fota_fmfu_dev_set_failed;
+
+	/* Expect other fakes to run successfully */
+	nct_init_fake.custom_fake = nct_init_success;
+	nfsm_init_fake.custom_fake = nfsm_init_success;
+	nrf_cloud_codec_init_fake.custom_fake = nrf_cloud_codec_init_success;
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = nrf_cloud_init(&params);
+
+	TEST_ASSERT_NOT_EQUAL_MESSAGE(0, ret,
+		"return should be non-zero when FOTA init failed");
+#else
+	TEST_IGNORE();
+#endif
+}
+
+/* TEST NFSM_SET_CURRENT_STATE_AND_NOTIFY */
+
+void test_set_current_state(void)
+{
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_TOTAL, nfsm_get_current_state(),
+		"nrf_cloud lib should be in STATE_TOTAL after changing");
+}
+
+/* TEST NRF_CLOUD_UNINIT */
+
+void test_uninit_fota_active(void)
+{
+#if defined(CONFIG_NRF_CLOUD_FOTA)
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	nct_init_fake.custom_fake = nct_init_success;
+	nfsm_init_fake.custom_fake = nfsm_init_success;
+	nrf_cloud_codec_init_fake.custom_fake = nrf_cloud_codec_init_success;
+#if defined(CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE)
+	nrf_cloud_fota_fmfu_dev_set_fake.custom_fake = nrf_cloud_fota_fmfu_dev_set_success;
+#endif
+	nrf_cloud_init(&params);
+
+	/* Expect nrf_cloud_fota_uninit fake returns busy */
+	nrf_cloud_fota_uninit_fake.custom_fake = nrf_cloud_fota_uninit_busy;
+	nct_disconnect_fake.custom_fake = nct_disconnect_success;
+
+	int ret = nrf_cloud_uninit();
+
+	TEST_ASSERT_EQUAL_MESSAGE(-EBUSY, ret,
+		"return should be -EBUSY when a FOTA job is active");
+#else
+	TEST_IGNORE();
+#endif
+}
+
+/* This test is ignored when the connection poll thread is active
+ * since it will triggered the transport disconnect event and give
+ * the uninit_disconnect semaphore which fails the test eventually
+ */
+void test_uninit_no_disconnect(void)
+{
+#if defined(CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD)
+	TEST_IGNORE();
+#else
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	nct_init_fake.custom_fake = nct_init_success;
+	nfsm_init_fake.custom_fake = nfsm_init_success;
+	nrf_cloud_codec_init_fake.custom_fake = nrf_cloud_codec_init_success;
+#if defined(CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE)
```

Comment:
Conditional behavior in unit tests should really be avoided if at all possible

Is it critical that we test all of these functions both with and without FMFU? Or can FMFU be separated into its own reduced-coverage test set? (Maybe one that re-uses your fakes, once they are separated into `nrf_cloud_fakes.h`)

For instance, I don't think there is any need to specifically test `uninit_no_disconnect` both with and without FMFU enabled.

I recommend instead that you test `uninit_no_disconnect` without FMFU, and in a separate FMFU test, briefly test that normal init/uninit work.